File: searchlib/src/vespa/searchlib/expression/aggregationrefnode.cpp
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
#include "aggregationrefnode.h"
#include <vespa/vespalib/util/stringfmt.h>
#include <stdexcept>

namespace search {
namespace expression {

using namespace vespalib;

IMPLEMENT_EXPRESSIONNODE(AggregationRefNode, ExpressionNode);

AggregationRefNode::AggregationRefNode(const AggregationRefNode & rhs) :
    ExpressionNode(),
    _index(rhs._index),
    _expressionNode(nullptr)
{
}

AggregationRefNode & AggregationRefNode::operator = (const AggregationRefNode & expr)
{
    if (this != &expr) {
        _index = expr._index;
        _expressionNode = nullptr;
    }
    return *this;
}

bool AggregationRefNode::onExecute() const
{
    if (_expressionNode != nullptr) {
        return _expressionNode->execute();
    }
    return false;
}

void AggregationRefNode::locateExpression(ExpressionNodeArray & exprVec) const
{
    if (_expressionNode == nullptr) {
        _expressionNode = static_cast<ExpressionNode *>(exprVec[_index].get());
        if (_expressionNode == nullptr) {
            throw std::runtime_error(make_string("Failed locating expression for index '%d'", _index));
        }
    }
}

Serializer & AggregationRefNode::onSerialize(Serializer & os) const
{
    return os << _index;
}

Deserializer & AggregationRefNode::onDeserialize(Deserializer & is)
{
    return is >> _index;
}

void
AggregationRefNode::visitMembers(vespalib::ObjectVisitor &visitor) const
{
    visit(visitor, "index", _index);
}

}
}

// this function was added by ../../forcelink.sh
void forcelink_file_searchlib_expression_expressionrefnode() {}

File: searchlib/src/main/java/com/yahoo/searchlib/expression/AggregationRefNode.java
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.searchlib.expression;

import com.yahoo.searchlib.aggregation.AggregationResult;
import com.yahoo.vespa.objects.Deserializer;
import com.yahoo.vespa.objects.ObjectVisitor;
import com.yahoo.vespa.objects.Serializer;

/**
 * This node holds the index of an ExpressionNode in an external array, and is used as a proxy in the back-end to allow
 * aggregators to be used in expressions.
 *
 * @author baldersheim
 */
public class AggregationRefNode extends ExpressionNode {

    public static final int classId = registerClass(0x4000 + 142, AggregationRefNode.class, AggregationRefNode::new);
    private AggregationResult result = null;
    private int index = - 1;

    public AggregationRefNode() { }

    public AggregationRefNode(int index) {
        this.index = index;
    }

    public AggregationRefNode(AggregationResult result) {
        this.result = result;
    }

    public AggregationResult getExpression() {
        return result;
    }

    public AggregationRefNode setExpression(AggregationResult result) {
        this.result = result;
        return this;
    }

    public AggregationRefNode setIndex(int index) {
        this.index = index;
        return this;
    }

    public int getIndex() {
        return index;
    }

    @Override
    public boolean onExecute() {
        return result.execute();
    }

    @Override
    public void onPrepare() {
        result.prepare();
    }

    @Override
    protected int onGetClassId() {
        return classId;
    }

    @Override
    protected void onSerialize(Serializer buf) {
        super.onSerialize(buf);
        buf.putInt(null, index);
    }

    @Override
    protected void onDeserialize(Deserializer buf) {
        super.onDeserialize(buf);
        index = buf.getInt(null);
        result = null;
    }

    @Override
    public AggregationRefNode clone() {
        AggregationRefNode obj = (AggregationRefNode)super.clone();
        obj.index = this.index;
        obj.result = this.result.clone();
        return obj;
    }

    @Override
    public void visitMembers(ObjectVisitor visitor) {
        super.visitMembers(visitor);
        visitor.visit("index", index);
    }

    @Override
    public ResultNode getResult() {
        return result.getResult();
    }

    @Override
    public int hashCode() {
        return super.hashCode() + index;
    }

    @Override
    public boolean equalsExpression(ExpressionNode obj) {
        AggregationRefNode rhs = (AggregationRefNode)obj;
        if (index != rhs.index) {
            return false;
        }
        if (!equals(result, rhs.result)) {
            return false;
        }
        return true;
    }
}
